### ISSUE TITLE
Genesis init bug

### ISSUE THREAD

**Issue by eyusufatik** [chainwayxyz/citrea#287]
## Summary

If no blocks are ran after genesis and node/sequencer is closed and reopened, genesis root is different.

## Steps to reproduce

1. Open sequencer
2. Don't publish any blocks
3. Close down sequencer
4. Reopen sequencer
5. Roots are different

Can ask @ercecan abotu this also, he found out about it

**Comment by eyusufatik** [comment]
One more bug with the storage system can be produced this way:

- Run sequencer
- Don't publish blocks
- Call `eth_getBalance` on any address in `genesis/evm.json`
- No balance
- Publish block
- Balance is correctly shown this time

Somehow storage right after genesis cannot be served from RPC

**Comment by eyusufatik** [comment]
Some takeaways from me (especially around the second comment):

In `crates/sovereign-sdk/module-system/sov-modules-rollup-blueprint/src/lib.rs` in functions `create_new_sequencer` and `create_new_rollup`, the `rpc_methods` variable is created with a reference to the storage before the chain is actually initialized. (chain is initialized in `CitreaSequencer::new` or `StateTransitionRunner::new`)
 
I suspect this would cause the problem (rpc's access to storage is behind `CitreaSequencer` or `StateTransitionRunner`'s access until a new version of the storage is created (by a new soft confirmation block)

My next step would be trying out calling `stf.init_chain()`inside rollup blueprint, before rpc is initialized instead of inside `CitreaSequencer` and `StateTransitionRunner`. But I don't know if it'll work.

**Comment by rakanalh** [comment]
I think the second case was indirectly fixed:
```
citrea on  nightly [$?] is 📦 v0.3.0 via 🦀 v1.77.1 took 7m30s
❯ ./target/debug/citrea --da-layer mock --rollup-config-path bin/citrea/configs/mock/sequencer_rollup_config.toml --sequencer-config-path bin/citrea/configs/mock/sequencer_config.toml --genesis-paths bin/test-data/genesis/demo-tests/mock
2024-04-06T20:48:41.603409Z DEBUG sov_stf_runner::config: Config file size: 43 bytes
2024-04-06T20:48:41.603676Z DEBUG sov_stf_runner::config: Config file size: 850 bytes
2024-04-06T20:48:41.625729Z DEBUG sov_mock_da::db_connector: Using test db: /home/rakan/Work/Chainway/citrea/test-da-dbs/main.db
2024-04-06T20:48:41.794578Z  INFO sov_schema_db: Opened RocksDB. rocksdb_name="ledger-db"
2024-04-06T20:48:41.850504Z  INFO sov_schema_db: Opened RocksDB. rocksdb_name="state-db"
2024-04-06T20:48:41.877808Z  INFO sov_schema_db: Opened RocksDB. rocksdb_name="native"
2024-04-06T20:48:41.877913Z DEBUG sov_prover_storage_manager: Giving 'finalized' storage ref with id 1
2024-04-06T20:48:41.882445Z  INFO ethereum_rpc: Failed to get latest git tag: ErrorObject { code: ServerError(-32001), message: "Failed to get version", data: Some(RawValue("Failure")) }
2024-04-06T20:48:41.883095Z  INFO citrea_sequencer::sequencer: No history detected. Initializing chain on block_header=MockBlockHeader { prev_hash: 0x0000000000000000000000000000000000000000000000000000000000000000, hash: 0x0101010101010101010101010101010101010101010101010101010101010101, txs_commitment: 0x0101010101010101010101010101010101010101010101010101010101010101, height: 0, time: Time { secs: 1672531200, nanos: 0 } }...
2024-04-06T20:48:41.883135Z DEBUG sov_prover_storage_manager: Requested native storage for block at height: 0, giving snapshot id=2
2024-04-06T20:48:41.883620Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14\x0f\x82\x0fB\x8a\xe46\xc1\0\x0b'W{\xf5\xbb\xf0\x9b\xfe\xc8\xf2" key_hash=KeyHash("00e99d6d518a8d79ce0fa77967ca81c524ee46767eec526cbd6f57d0c05c1340")
2024-04-06T20:48:41.883691Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14:\xee\xb8q\xf8<\x85\xe6\x8f\xfd\x18h\xbe\xf3B^\xd6d\x9d9" key_hash=KeyHash("39871c5e6b3e83a75ef5ff4c5357f53ddf8e1430066617d9b4df71b2b84dc089")
2024-04-06T20:48:41.883747Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14f\xf6\x86\x92\xc0>\xb9\xc0emgo/K\xd1>\xba@\xd1\xb7" key_hash=KeyHash("d125708efbbf753ba97e3f0ef2637ed5c8b7067771868eb3e5d874b5e21a8999")
2024-04-06T20:48:41.883793Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14\x9f\xcd\xf8\xf6\r0\tenP\xbf\x80\\\xd5<s5\xb2\x84\xfb" key_hash=KeyHash("2e14236798906d85bef1d044c7f9b844ac10e36fc9699eb0cb8f6d1d9034717a")
2024-04-06T20:48:41.883833Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14\xaa\xfbtB\xf7\xf0\x0bd\x05|.\x9e\xae(\x15\xbbc\xee\x0e\xce" key_hash=KeyHash("f4361f91140a7a2d19921338a509335c9ba734c29e80801268cc5098e13eaeca")
2024-04-06T20:48:41.883881Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14\xc2\xf8\xee\xd7}\xa1X?{\xae\n1%\xdc{\xc4&\0-\xde" key_hash=KeyHash("082e861608ba978f6a8b3e3fcf81ba2fc66747c75bf857d0c0677d6d852996f6")
2024-04-06T20:48:41.883942Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14\xd4H!\xf9\x06\xe3\x90\x9b\x8a\xe9D\xf7\x06\x05Q\xc3;\x92,\xc9" key_hash=KeyHash("4e18fdfe75094c7630094556c90085c4e361360cfc3c8981943906390483166d")
2024-04-06T20:48:41.883993Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14\xe7V\xfd\xf8\x93g\xefB\x8bH\xbc\xa2\xd2r\xec\x8e\xce\xc0S\xfd" key_hash=KeyHash("3b9f608d673c1e7b3ffcdbf7b3aaf90511a8d8ca18a4fb0285984491a6e4c623")
2024-04-06T20:48:41.884061Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14\xf3\x9f\xd6\xe5\x1a\xad\x88\xf6\xf4\xcej\xb8\x82ry\xcf\xff\xb9\"f" key_hash=KeyHash("a3137c7735d1989855cf5b93419289a3c55d8d4a9d80c63ea3375f7f29e0da27")
2024-04-06T20:48:41.884107Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/cfg/" key_hash=KeyHash("a55523b1752e8098bea68ccfb385e0ca2e83378ba2b364cef6a194267677285e")
2024-04-06T20:48:41.884163Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/head/" key_hash=KeyHash("2586d4fa8e61a0ab5b3bf55417d3272dda73f8e5e37aa0654476ca1836111b3a")
2024-04-06T20:48:41.884240Z DEBUG jmt: hashed jmt key key=b"soft_confirmation_rule_enforcer/SoftConfirmationRuleEnforcer/authority/" key_hash=KeyHash("a3eaf7514b8c3f8ddf3501a1435ad0e122652cc411292fa28f46221bfb984531")
2024-04-06T20:48:41.884320Z DEBUG jmt: hashed jmt key key=b"soft_confirmation_rule_enforcer/SoftConfirmationRuleEnforcer/l1_fee_rate_change_percentage/" key_hash=KeyHash("071ff0a2c2e065e73c1837d6a74ddebedd6070fdb3617930f14338eef3320e78")
2024-04-06T20:48:41.884377Z DEBUG jmt: hashed jmt key key=b"soft_confirmation_rule_enforcer/SoftConfirmationRuleEnforcer/limiting_number/" key_hash=KeyHash("144164068dbd8323788fcb585f8f7c39c07bd792d460c0b5200ff4514333c6d7")
2024-04-06T20:48:41.884423Z DEBUG jmt: hashed jmt key key=b"sov_chain_state/ChainState/genesis_height/" key_hash=KeyHash("f9e03d7a524c1b8cc55d504629ae936e9e688c321f8a5a2f60188b3d2802e170")
2024-04-06T20:48:41.884483Z DEBUG jmt: hashed jmt key key=b"sov_chain_state/ChainState/time/\0\0\0\0\0\0\0\0" key_hash=KeyHash("2b314206abae08fb880151ae7c442fcf3a8515d93e37743d815b770b5ea1ab7a")
2024-04-06T20:48:41.884551Z DEBUG jmt: hashed jmt key key=b"sov_chain_state/ChainState/true_height/" key_hash=KeyHash("965364ea40b78916fd31030663374b6e618ef83263e738351493847ebe137811")
2024-04-06T20:48:41.885144Z DEBUG sov_prover_storage_manager: Snapshot id=2 for block at height=0 has been saved to StorageManager
2024-04-06T20:48:41.888004Z  INFO citrea_sequencer::sequencer: Chain initialization is done. Genesis root: 0x0d3206ffe071ad155f517a2574951f922a29ddbda954733fa2b4d4b3f2238d69
2024-04-06T20:48:41.888058Z  INFO citrea_evm::query: evm module: eth_getBlockByNumber
2024-04-06T20:48:41.963055Z  INFO citrea_sequencer::sequencer: Starting RPC server at 127.0.0.1:12345
2024-04-06T20:48:41.969544Z DEBUG sov_mock_da::service: Finalized MockHeader: MockBlockHeader { height: 1, prev_hash: 0101010101010101010101010101010101010101010101010101010101010101, next_hash: 1e9dbc1a11f8e046a72d1296cc2d8bb0d1544d56fd0b9bb8890a0f89b8803654 }
^C

citrea on  nightly [$?] is 📦 v0.3.0 via 🦀 v1.77.1 took 5s
❯ ./target/debug/citrea --da-layer mock --rollup-config-path bin/citrea/configs/mock/sequencer_rollup_config.toml --sequencer-config-path bin/citrea/configs/mock/sequencer_config.toml --genesis-paths bin/test-data/genesis/demo-tests/mock
2024-04-06T20:48:47.796539Z DEBUG sov_stf_runner::config: Config file size: 43 bytes
2024-04-06T20:48:47.796930Z DEBUG sov_stf_runner::config: Config file size: 850 bytes
2024-04-06T20:48:47.818625Z DEBUG sov_mock_da::db_connector: Using test db: /home/rakan/Work/Chainway/citrea/test-da-dbs/main.db
2024-04-06T20:48:47.841097Z  INFO sov_schema_db: Opened RocksDB. rocksdb_name="ledger-db"
2024-04-06T20:48:47.856827Z  INFO sov_schema_db: Opened RocksDB. rocksdb_name="state-db"
2024-04-06T20:48:47.867996Z  INFO sov_schema_db: Opened RocksDB. rocksdb_name="native"
2024-04-06T20:48:47.868056Z DEBUG sov_prover_storage_manager: Giving 'finalized' storage ref with id 1
2024-04-06T20:48:47.871825Z  INFO ethereum_rpc: Failed to get latest git tag: ErrorObject { code: ServerError(-32001), message: "Failed to get version", data: Some(RawValue("Failure")) }
2024-04-06T20:48:47.872529Z DEBUG citrea_sequencer::sequencer: Chain is already initialized. Skipping initialization.
2024-04-06T20:48:47.872617Z  INFO citrea_evm::query: evm module: eth_getBlockByNumber
2024-04-06T20:48:47.954158Z  INFO citrea_sequencer::sequencer: Starting RPC server at 127.0.0.1:12345
2024-04-06T20:49:09.757613Z DEBUG connection{remote_addr=127.0.0.1:48956 conn_id=0}: jsonrpsee_server::server: Accepting new connection 1/10000
2024-04-06T20:49:09.760018Z  INFO citrea_evm::query: evm module: eth_getBalance
```

```
citrea on  nightly [$?] is 📦 v0.3.0 via 🦀 v1.77.1
❯ cast balance --rpc-url http://localhost:12345 0xC2F8Eed77da1583f7bae0a3125Dc7BC426002dDE
1329227995784915872903807060280344575
```

**Comment by rakanalh** [comment]
I can confirm that the first case applies.

After running a new sequencer node at genesis, the `prover_storage` returns a root of: `0x5350415253455f4d45524b4c455f504c414345484f4c4445525f484153485f5f` while AFTER this value is generated, the genesis root calculated when the initial db state is geenrated is: `0x0d3206ffe071ad155f517a2574951f922a29ddbda954733fa2b4d4b3f2238d69`

**Comment by rakanalh** [comment]
Could not reproduce second case:

```
citrea on  rakan/genesis-bug-fix [$?] is 📦 v0.3.0 via 🦀 v1.77.1 took 14s
❯ make clean-node
rm -rf sequencer-db
rm -rf full-node-db
rm test-da-dbs/*.db

citrea on  rakan/genesis-bug-fix [$?] is 📦 v0.3.0 via 🦀 v1.77.1
❯ RUST_BACKTRACE=1 ./target/debug/citrea --da-layer mock --rollup-config-path bin/citrea/configs/mock/sequencer_rollup_config.toml --sequencer-config-path bin/citrea/configs/mock/sequencer_config.toml --genesis-paths bin/test-data/genesis/demo-tests/mock
2024-04-08T01:45:43.594430Z DEBUG sov_stf_runner::config: Config file size: 43 bytes
2024-04-08T01:45:43.594667Z DEBUG sov_stf_runner::config: Config file size: 850 bytes
2024-04-08T01:45:43.612853Z DEBUG sov_mock_da::db_connector: Using test db: /home/rakan/Work/Chainway/citrea/test-da-dbs/main.db
2024-04-08T01:45:43.749425Z  INFO sov_schema_db: Opened RocksDB. rocksdb_name="ledger-db"
2024-04-08T01:45:43.785221Z  INFO sov_schema_db: Opened RocksDB. rocksdb_name="state-db"
2024-04-08T01:45:43.807088Z  INFO sov_schema_db: Opened RocksDB. rocksdb_name="native"
2024-04-08T01:45:43.807215Z DEBUG sov_prover_storage_manager: Giving 'finalized' storage ref with id 1
2024-04-08T01:45:43.810994Z  INFO ethereum_rpc: Failed to get latest git tag: ErrorObject { code: ServerError(-32001), message: "Failed to get version", data: Some(RawValue("Failure")) }
2024-04-08T01:45:43.811773Z  INFO citrea_sequencer::sequencer: No history detected. Initializing chain on block_header=MockBlockHeader { prev_hash: 0x0000000000000000000000000000000000000000000000000000000000000000, hash: 0x0101010101010101010101010101010101010101010101010101010101010101, txs_commitment: 0x0101010101010101010101010101010101010101010101010101010101010101, height: 0, time: Time { secs: 1672531200, nanos: 0 } }...
2024-04-08T01:45:43.811827Z DEBUG sov_prover_storage_manager: Requested native storage for block at height: 0, giving snapshot id=2
2024-04-08T01:45:43.812549Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14\x0f\x82\x0fB\x8a\xe46\xc1\0\x0b'W{\xf5\xbb\xf0\x9b\xfe\xc8\xf2" key_hash=KeyHash("00e99d6d518a8d79ce0fa77967ca81c524ee46767eec526cbd6f57d0c05c1340")
2024-04-08T01:45:43.812688Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14:\xee\xb8q\xf8<\x85\xe6\x8f\xfd\x18h\xbe\xf3B^\xd6d\x9d9" key_hash=KeyHash("39871c5e6b3e83a75ef5ff4c5357f53ddf8e1430066617d9b4df71b2b84dc089")
2024-04-08T01:45:43.812804Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14f\xf6\x86\x92\xc0>\xb9\xc0emgo/K\xd1>\xba@\xd1\xb7" key_hash=KeyHash("d125708efbbf753ba97e3f0ef2637ed5c8b7067771868eb3e5d874b5e21a8999")
2024-04-08T01:45:43.812896Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14\x9f\xcd\xf8\xf6\r0\tenP\xbf\x80\\\xd5<s5\xb2\x84\xfb" key_hash=KeyHash("2e14236798906d85bef1d044c7f9b844ac10e36fc9699eb0cb8f6d1d9034717a")
2024-04-08T01:45:43.812988Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14\xaa\xfbtB\xf7\xf0\x0bd\x05|.\x9e\xae(\x15\xbbc\xee\x0e\xce" key_hash=KeyHash("f4361f91140a7a2d19921338a509335c9ba734c29e80801268cc5098e13eaeca")
2024-04-08T01:45:43.813100Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14\xc2\xf8\xee\xd7}\xa1X?{\xae\n1%\xdc{\xc4&\0-\xde" key_hash=KeyHash("082e861608ba978f6a8b3e3fcf81ba2fc66747c75bf857d0c0677d6d852996f6")
2024-04-08T01:45:43.813239Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14\xd4H!\xf9\x06\xe3\x90\x9b\x8a\xe9D\xf7\x06\x05Q\xc3;\x92,\xc9" key_hash=KeyHash("4e18fdfe75094c7630094556c90085c4e361360cfc3c8981943906390483166d")
2024-04-08T01:45:43.813335Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14\xe7V\xfd\xf8\x93g\xefB\x8bH\xbc\xa2\xd2r\xec\x8e\xce\xc0S\xfd" key_hash=KeyHash("3b9f608d673c1e7b3ffcdbf7b3aaf90511a8d8ca18a4fb0285984491a6e4c623")
2024-04-08T01:45:43.813484Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/accounts/\x14\xf3\x9f\xd6\xe5\x1a\xad\x88\xf6\xf4\xcej\xb8\x82ry\xcf\xff\xb9\"f" key_hash=KeyHash("a3137c7735d1989855cf5b93419289a3c55d8d4a9d80c63ea3375f7f29e0da27")
2024-04-08T01:45:43.813590Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/cfg/" key_hash=KeyHash("a55523b1752e8098bea68ccfb385e0ca2e83378ba2b364cef6a194267677285e")
2024-04-08T01:45:43.813716Z DEBUG jmt: hashed jmt key key=b"citrea_evm/Evm/head/" key_hash=KeyHash("2586d4fa8e61a0ab5b3bf55417d3272dda73f8e5e37aa0654476ca1836111b3a")
2024-04-08T01:45:43.813880Z DEBUG jmt: hashed jmt key key=b"soft_confirmation_rule_enforcer/SoftConfirmationRuleEnforcer/authority/" key_hash=KeyHash("a3eaf7514b8c3f8ddf3501a1435ad0e122652cc411292fa28f46221bfb984531")
2024-04-08T01:45:43.814046Z DEBUG jmt: hashed jmt key key=b"soft_confirmation_rule_enforcer/SoftConfirmationRuleEnforcer/l1_fee_rate_change_percentage/" key_hash=KeyHash("071ff0a2c2e065e73c1837d6a74ddebedd6070fdb3617930f14338eef3320e78")
2024-04-08T01:45:43.814168Z DEBUG jmt: hashed jmt key key=b"soft_confirmation_rule_enforcer/SoftConfirmationRuleEnforcer/limiting_number/" key_hash=KeyHash("144164068dbd8323788fcb585f8f7c39c07bd792d460c0b5200ff4514333c6d7")
2024-04-08T01:45:43.814273Z DEBUG jmt: hashed jmt key key=b"sov_chain_state/ChainState/genesis_height/" key_hash=KeyHash("f9e03d7a524c1b8cc55d504629ae936e9e688c321f8a5a2f60188b3d2802e170")
2024-04-08T01:45:43.814401Z DEBUG jmt: hashed jmt key key=b"sov_chain_state/ChainState/time/\0\0\0\0\0\0\0\0" key_hash=KeyHash("2b314206abae08fb880151ae7c442fcf3a8515d93e37743d815b770b5ea1ab7a")
2024-04-08T01:45:43.814611Z DEBUG jmt: hashed jmt key key=b"sov_chain_state/ChainState/true_height/" key_hash=KeyHash("965364ea40b78916fd31030663374b6e618ef83263e738351493847ebe137811")
2024-04-08T01:45:43.815476Z DEBUG sov_prover_storage_manager: Snapshot id=2 for block at height=0 has been saved to StorageManager
2024-04-08T01:45:43.822110Z  INFO citrea_sequencer::sequencer: Chain initialization is done. Genesis root: 0x0d3206ffe071ad155f517a2574951f922a29ddbda954733fa2b4d4b3f2238d69
2024-04-08T01:45:43.822200Z  INFO citrea_evm::query: evm module: eth_getBlockByNumber
2024-04-08T01:45:43.894357Z  INFO citrea_sequencer::sequencer: Starting RPC server at 127.0.0.1:12345
2024-04-08T01:45:43.897861Z DEBUG sov_mock_da::service: Finalized MockHeader: MockBlockHeader { height: 1, prev_hash: 0101010101010101010101010101010101010101010101010101010101010101, next_hash: 1e9dbc1a11f8e046a72d1296cc2d8bb0d1544d56fd0b9bb8890a0f89b8803654 }
2024-04-08T01:45:49.353457Z DEBUG connection{remote_addr=127.0.0.1:47216 conn_id=0}: jsonrpsee_server::server: Accepting new connection 1/10000
2024-04-08T01:45:49.354235Z  INFO citrea_evm::query: evm module: eth_getBalance
2024-04-08T01:45:51.805369Z DEBUG connection{remote_addr=127.0.0.1:41000 conn_id=1}: jsonrpsee_server::server: Accepting new connection 1/10000
2024-04-08T01:45:51.807669Z  INFO citrea_evm::query: evm module: eth_getBalance
2024-04-08T01:45:54.288656Z DEBUG connection{remote_addr=127.0.0.1:41006 conn_id=2}: jsonrpsee_server::server: Accepting new connection 1/10000
2024-04-08T01:45:54.289639Z  INFO citrea_evm::query: evm module: eth_getBalance
```

The node was up and running, querying balances:
```
citrea on  rakan/genesis-bug-fix [$?] is 📦 v0.3.0 via 🦀 v1.77.1
❯ cast balance --rpc-url http://localhost:12345 0xd44821f906E3909b8AE944F7060551c33b922cc9
1329227995784915872903807060280344575

citrea on  rakan/genesis-bug-fix [$?] is 📦 v0.3.0 via 🦀 v1.77.1
❯ cast balance --rpc-url http://localhost:12345 0x0f820f428AE436C1000b27577bF5bbf09BfeC8f2
1329227995784915872903807060280344575

citrea on  rakan/genesis-bug-fix [$?] is 📦 v0.3.0 via 🦀 v1.77.1
❯ cast balance --rpc-url http://localhost:12345 0xC2F8Eed77da1583f7bae0a3125Dc7BC426002dDE
1329227995784915872903807060280344575
```
